Fix issue 7280 - issue warning after retry

danizen opened the pull request:
- Adds a new test utility in `tests\lib\fileystem.py` designed to help create multi-process race conditions of this kind.  Tests for test utility are in `tests\unit\test_utils_filesystem.py`
- Tests the race condition with code in `tests\unit\test_util_temp_dir.py`
- Address the issue by catching the error after retry in `src\pip\_internal\util\temp_dir.py` and issuing a warning.

danizen commented:
@pfmoore @chrahunt, I am completely confused about the appveyor failure.  This status response from psutil isn't one of the ones present in psutil statuses, and ignoring by adding something like this didn't seem to show the latest code:

```python
latest_children = [child for child in process.children() if child.status() != 'terminated']
```

It does seem to me to indicate that the exception for a missing file is not properly caught and printed, but I have not been able to reproduce that on my own.

chrahunt commented on tests/unit/test_utils_filesystem.py:
```diff
@@ -59,3 +70,53 @@ def test_copy2_fixed_raises_appropriate_errors(create, error_type, tmpdir):
+def test_file_opener_no_file(process):
+    # FileOpener joins the subprocess even if the parent never sends the path
+    with FileOpener():
+        pass
```
We can also assert that we have children in the `with` statement, otherwise this may pass incorrectly if we make a change that never starts the child process in the first place.

chrahunt commented on tests/unit/test_utils_filesystem.py:
```diff
@@ -59,3 +70,53 @@ def test_copy2_fixed_raises_appropriate_errors(create, error_type, tmpdir):
+    # The FileOpener cleans up the subprocess when the file cannot be opened
+    path = tmpdir.joinpath('foo.txt')
+    with FileOpener(path):
+        pass
```
Similar to above, we should assert that we have a child process here.

danizen replied:
Done

chrahunt commented on tests/unit/test_utils_filesystem.py:
```diff
@@ -59,3 +70,53 @@ def test_copy2_fixed_raises_appropriate_errors(create, error_type, tmpdir):
+        f.write('Hello\n')
+    with FileOpener() as opener:
+        opener.send(path)
+        with pytest.raises(OSError):
```
Can we assert facts about this error?

danizen replied:
Yes, we can insist it is a EACCES error.  Part of the point of my patch to pip/_internal/utils/temp_dir.py, is that it doesn't matter whether it is in fact windows error 5 or 32, because we know this is a temporary directory we just created.

danizen replied:
Done

chrahunt commented on tests/unit/test_utils_filesystem.py:
```diff
@@ -59,3 +70,53 @@ def test_copy2_fixed_raises_appropriate_errors(create, error_type, tmpdir):
+    os.mkdir(subdir)
+    path = subdir.joinpath('bar.txt')
+    with open(path, 'w') as f:
+        f.write('Hello\n')
```
This can be shortened to `path.write_text("Hello")`.

danizen replied:
I can make this change.  I'd missed the difference between `tmp_path` and `tmpdir` in pytest, and was concerned about back-port to Python 2.7.

chrahunt replied:
We override `tmpdir` in our conftest to return a `tests.lib.path.Path` (which is not at all obvious or easily discovered).

danizen replied:
Done

chrahunt commented on tests/unit/test_utils_filesystem.py:
```diff
@@ -59,3 +70,53 @@ def test_copy2_fixed_raises_appropriate_errors(create, error_type, tmpdir):
     assert not dest.exists()
+
+
+def test_file_opener_no_file(process):
```
These `FileOpener` tests would be better in `tests/lib/test_filesystem.py`. This file is really for tests of functionality in `pip/_internal/utils/filesystem.py`.

chrahunt commented on tests/unit/test_utils_filesystem.py:
```diff
@@ -59,3 +70,53 @@ def test_copy2_fixed_raises_appropriate_errors(create, error_type, tmpdir):
+
+
+@skip_unless_windows
+def test_file_opener_produces_unlink_error(tmpdir, process):
```
Since these tests are more against Python/Windows than anything in `pip/_internal/utils/filesystem.py`, I would put them in `tests/functional/test_windows.py`.

danizen replied:
I think moving them to `tests/lib/test_lib_filesystem.py` is hopefully enough. Since producing these errors was the whole point, I'd rather not separate these tests from the other tests.   Please resolve if you think this is OK.

danizen commented:
@chrahunt, I've done some of the work to make this pull request only about the new tests, rather than about issue #7280, with the following exceptions:

* the new unit test in "tests/unit/tests_util_temp_dir.py" is marked as `pytest.mark.xfail` - so, I've produced a failing test for TempDirectory that illustrates the issue we are trying to address
* there is still news/7280.bugfix, because I haven't quite despaired of addressing the actual issue with this pull request.  However, I want to be clearer that the FileOpener does reproduce the issue.

I think I've addressed the issues with the pull you raised.   I named the test file as tests/lib/test_lib_filesystem, just because that is clearer to me when I enter pytest -k test_lib, I get all the tests on tests/lib.  I am very open to changing that; it just seemed a small bit better to me.

danizen commented on tests/lib/filesystem.py:
```diff
@@ -34,6 +36,101 @@ def make_unreadable_file(path):
+            # Now path is open and we wait for signal to exit
+            conn.recv()
+    finally:
+        if f:
```
This suggested change did not work for me, and caused the sub-process or main process to hang.  In general, I liked it more that the sub-process was using primitive parts of Python, and was independent of pip's own infrastructure, so I thought I'd raise that as a good thing before we potentially figure out what I did wrong :)

danizen commented on tests/lib/filesystem.py:
```diff
@@ -34,6 +36,101 @@ def make_unreadable_file(path):
+            # Open the file
+            try:
+                f = open(path, 'r')
+                # NOTE: action is for future use and may be unused
```
Removed the action

danizen commented on tests/lib/filesystem.py:
```diff
@@ -34,6 +36,101 @@ def make_unreadable_file(path):
+            self.send(path, action)
+
+    def send(self, path, action=None):
+        if self.path is not None:
```
This was done

danizen commented:
I have pull request fatigue.  All of this works on my Desktop, both as davisda4 and as a local administrator.  I've run tox -e py35, tox -e py36, and tox -e py27.

danizen commented:
> I have pull request fatigue. All of this works on my Desktop, both as davisda4 and as a local administrator. I've run tox -e py35, tox -e py36, and tox -e py27.

I spoke too quickly. I have run these specific unit tests, but I am now re-running the entire suite with Python 3.6 on Windows, and we'll see.

pradyunsg commented:
Hi @danizen! Is there anything we can help you with to move this PR forward?

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will be eligible for code review and hopefully merging!

danizen commented:
> Hi @danizen! Is there anything we can help you with to move this PR forward?

@pradyunsg,

I was mostly interested in issue #7280, which we are no longer seeing since pip 20.  I think the main work I and the other contributor did is to clarify the issue, but moving the TemporaryDirectory clean-up for the wheel to the end (by making it global), ended up fixing the issue.

As a contractor, I've pushed my government boss as far as I can to get her to let me do open source contributions. The agreement we worked out was that submitting issues and clarifying them is something they'll pay for, but fixing them I'm on my own.  So, I think I'm done.  There is no reason I would want to make Windows easier and more foolproof except for my employer preferring Windows development boxes.

SanketDG commented:
Isn't this largely superseded (especially during installing) with https://github.com/pypa/pip/pull/7882?

danizen commented:
> Isn't this largely superseded (especially during installing) with #7882?

@SanketDG,

Looks like - in any case, this pull request is no longer associated with issue #7280, as I was asked to make it only test code.
